webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/servers/RemoveNodeService.scala

```diff
@@ -178,105 +233,63 @@ class RemoveNodeServiceImpl(
-                             case Full(x) => Some(x)
-                             case eb:EmptyBox => // if the policy server is not found, we must not fails (#11231)
-                               val msg = (eb ?~! s"Error when trying to calculate node '${nodeId.value}' policy path").messageChain
-                               logEffect.warn(msg)
-                               None
-                           }
-          startPreHooks =  System.currentTimeMillis
-          preHooks      <- RunHooks.getHooks(HOOKS_D + "/node-pre-deletion", HOOKS_IGNORE_SUFFIXES)
-          result        <- effectiveDeletion(nodeInfo, nodePaths, preHooks, startPreHooks)
+          _      <- NodeLoggerPure.Delete.debug(s"Deleting node with ID '${nodeId.value}' [mode:${mode.name}]")
+          status <- findNodeStatus(nodeId)
+          -      <- NodeLoggerPure.Delete.debug(s"  - node '${nodeId.value}' has status: [${status.map(_.name).mkString(",")}]")
+          info   <- Ref.make(Option.empty[NodeInfo]) //a place to store the maybe node info
+          // always delete in order pending then accepted then deleted
+           res1  <- if(status.contains(PendingInventory)) {
+                      (for {
+                        r <- deletePendingNode(nodeId, mode, modId, actor)
+                        i <- nodeInfoService.getPendingNodeInfoPure(nodeId)
+                        _ <- info.set(i)
+                      } yield r).catchAll(err => Error(err).succeed)
+                    } else Success.succeed
+           res2  <- if(status.contains(AcceptedInventory)) {
+                      (for {
+                        i <- nodeInfoService.getNodeInfoPure(nodeId)
+                        r <- i match {
+                               case None    => Success.succeed // perhaps deleted or something
+                               case Some(x) => info.set(Some(x)) *> deleteAcceptedNode(x, mode, modId, actor)
+                             }
+                      } yield r).catchAll(err => Error(err).succeed)
+                    } else Success.succeed
+           res3  <- if(status.contains(RemovedInventory)) {
+                      (for {
+                        i <- nodeInfoService.getDeletedNodeInfoPure(nodeId)
+                        r <- deleteDeletedNode(nodeId, mode, modId, actor)
+                        // only update if nodeInfo is not already set, b/c accepted has more info
+                        _ <- info.update(opt => opt.orElse(i))
+                      } yield r).catchAll(err => Error(err).succeed)
+                    } else Success.succeed
+           // if any of the previous cases were in error, we want to stop here
+           res    <- DeletionResult.merge(res1, res2 :: res3 :: Nil)
+           // in all cases, run postNodeDeletionAction
+          _       <- NodeLoggerPure.Delete.debug(s"-> execute clean-up actions for node '${nodeId.value}'")
+          actions <- postNodeDeleteActions.get
+          optInfo <- info.get
```

Comment:
Here you have only one info although you may have 3 nodes, and the "Removed" has priority over accepted and pending.

Maybe you should have a list more than a ref here, and you can explain more in details what has been removed

Reply:
I don't think so, for delete it's `info.update(opt => opt.orElse(i))`, so only if there was no other info previously set